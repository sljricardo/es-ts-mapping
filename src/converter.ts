function typeAs(value: string) {
    const toStringDataTypes = ["keyword", "text"]
    const toNumberDataTypes = ["float", "integer", "double", "long"]
    const toDateDataTypes = ["date"]
    const toBoleanDataTypes = ["boolean"]

    if (toStringDataTypes.includes(value)) return "string | string[]"
    if (toNumberDataTypes.includes(value)) return "number | number[]"
    if (toDateDataTypes.includes(value)) return "string | string[] | number | number[]"
    if (toBoleanDataTypes.includes(value)) return "boolean | boolean[]"
    return "any | any[]"
}

function innerMapping(object: { [key: string]: any }, count = 1) {

    let internal = ""

    Object.keys(object).forEach(key => {
        const identation = [...Array(count)].map(k => " ").join("")
        if (object[key]?.properties) {
            internal += `${identation}${key}: {\n${innerMapping(object[key]?.properties, count + 1)}${identation}},\n`
        } else {
            internal += `${identation}${key}: ${typeAs(object[key]?.type)}\n`
        }
    });

    return internal
}

export function convertToTypescript(object: { [key: string]: any }, count = 1) {
    let internal = ""

    Object.keys(object).forEach(key => {
        internal += count === 1 && `\nexport type ${key}ElasticType = {\n`

        if (Object.keys(object[key])[0] === "mappings") {
            internal += innerMapping(object[key].mappings.properties)
        }

        internal += count === 1 && `}\n`
    });

    return internal
}