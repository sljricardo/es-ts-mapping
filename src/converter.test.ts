import { expect, test } from 'vitest'
import { convertToTypescript } from './converter';

test('expect to get typescript object type', () => {
  const input = JSON.parse("{\"reviews\":{\"mappings\":{\"properties\":{\"author\":{\"properties\":{\"email\":{\"type\":\"keyword\"},\"first_name\":{\"type\":\"text\"},\"last_name\":{\"type\":\"text\"}}},\"content\":{\"type\":\"text\"},\"created_at\":{\"type\":\"date\"},\"date\":{\"type\":\"long\"},\"product_id\":{\"type\":\"integer\"},\"rating\":{\"type\":\"float\"},\"reviews\":{\"type\":\"nested\",\"properties\":{\"place\":{\"type\":\"text\",\"fields\":{\"keyword\":{\"type\":\"keyword\",\"ignore_above\":256}}}}}}}},\"products\":{\"mappings\":{\"properties\":{\"created\":{\"type\":\"date\",\"format\":\"yyyy/MM/ddHH:mm:ss||yyyy/MM/dd||epoch_millis\"},\"description\":{\"type\":\"text\",\"fields\":{\"keyword\":{\"type\":\"keyword\",\"ignore_above\":256}}},\"in_stock\":{\"type\":\"long\"},\"is_active\":{\"type\":\"boolean\"},\"name\":{\"type\":\"text\",\"fields\":{\"keyword\":{\"type\":\"keyword\",\"ignore_above\":256}}},\"price\":{\"type\":\"long\"},\"sold\":{\"type\":\"long\"},\"tags\":{\"type\":\"text\",\"fields\":{\"keyword\":{\"type\":\"keyword\",\"ignore_above\":256}}}}}}}")

  expect(convertToTypescript(input)).toMatchSnapshot()
});