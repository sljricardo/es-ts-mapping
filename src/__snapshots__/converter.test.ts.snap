// Vitest Snapshot v1

exports[`expect to get typescript object type 1`] = `
"
export type reviewsElasticType = {
 author: {
  email: string | string[]
  first_name: string | string[]
  last_name: string | string[]
 },
 content: string | string[]
 created_at: string | string[] | number | number[]
 date: number | number[]
 product_id: number | number[]
 rating: number | number[]
 reviews: {
  place: string | string[]
 },
}

export type productsElasticType = {
 created: string | string[] | number | number[]
 description: string | string[]
 in_stock: number | number[]
 is_active: boolean | boolean[]
 name: string | string[]
 price: number | number[]
 sold: number | number[]
 tags: string | string[]
}
"
`;
